File: go/pkg/2024/day2.go
package solver2024

import (
	"fmt"
	"strconv"
	"strings"
)

type SolverTwo struct {
}

func (solver SolverTwo) SolutionOne(lines []string) int {
    reports := makeReportsFromLines(lines)
    answer := 0

    for _, element := range reports {
        if isSafeReport(element.numbers) {
            answer++
        }
    }

    return answer;
}

func (solver SolverTwo) SolutionTwo(lines []string) int {
    reports := makeReportsFromLines(lines)
    answer := 0

    for _, element := range reports {
        for i := -1; i < len(element.numbers); i++ {
            list := element.numbers
            if i >= 0 {
                list = RemoveIndex(list, i)
            }
            fmt.Printf("List: %v\n", list)
            safe := isSafeReport(list)

            if safe {
                answer++
                break
            }
        }
    }

    return answer;
}

type report struct {
    numbers []int
}

func (report report) isSafe() bool {
    var previousNumber int
    var isAscending bool

    for index, element := range report.numbers {
        if index == 0 {
            previousNumber = element
            continue
        }

        if index == 1 {
            if element > previousNumber {
                isAscending = true
            } else {
                isAscending = false
            }
        }

        if directionChanged(previousNumber, element, isAscending) || previousNumber == element {
            return false
        }

        difference := element - previousNumber

        if isAscending && difference > 3 || !isAscending && difference < -3 {
            return false
        }

        previousNumber = element
    }

    return true
}

func isSafeReport(numbers []int) bool {
    // fmt.Printf("Checking %v\n", numbers)
    var previousNumber int
    var isAscending bool

    for index, element := range numbers {
        // fmt.Printf("Current index: %d, element: %d\n", index, element)
        if index == 0 {
            previousNumber = element
            continue
        }

        if index == 1 {
            if element > previousNumber {
                isAscending = true
            } else {
                isAscending = false
            }
        }

        if !isSafeSequence(previousNumber, element, isAscending) {
            // fmt.Println("Not safe")
            return false
        }

        previousNumber = element
    }

    return true
}

func isSafeSequence(previousNumber int, nextNumber int, isAscending bool) bool {
    if directionChanged(previousNumber, nextNumber, isAscending) {
        return false
    }
    if previousNumber == nextNumber {
        return false
    }

    difference := nextNumber - previousNumber

    if isAscending && difference > 3 || !isAscending && difference < -3 {
        return false
    }

    return true
}

func directionChanged(previousNumber int, nextNumber int, isAscending bool) bool {
    if isAscending && previousNumber > nextNumber {
        return true
    } else if !isAscending && previousNumber < nextNumber {
        return true
    }

    return false
}

func makeReportsFromLines(lines []string) []report {
    reports := []report{}

    for _, element := range lines {
        reports = append(reports, makeReportFromLine(element))
    }

    return reports
}

func makeReportFromLine(line string) report {
    numbers := strings.Split(line, " ")

    structNumbers := []int{}

    for _, number := range numbers {
        newNumber, _ := strconv.Atoi(number)
        structNumbers = append(structNumbers, newNumber)
    }

    return report{
        structNumbers,
    }
}

func RemoveIndex(s []int, index int) []int {
    newSlice := []int{}
    newSlice = append(newSlice, s[:index]...)
    newSlice = append(newSlice, s[index+1:]...)
    return newSlice
}

File: rust/aoc/src/solver/s24/day2.rs
use crate::solver::solver::Solver;

pub struct Day2 {
}

impl Solver for Day2 {
    fn solution_one(&self, lines: Vec<String>) -> i64 {
        let answer: i64 = 0;

        for line in lines {
            todo!();
        }

        return answer;
    }

    fn solution_two(&self, lines: Vec<String>) -> i64 {
        for line in lines {
            todo!();
        }

        return 0;
    }
}

fn is_safe_report(numbers: Vec<i64>) -> bool {
    let mut previous_number: i64;
    let mut is_ascending: bool;
    
    for (index, number) in numbers.iter().enumerate() {
        if index == 0 {
            previous_number = *number;
            continue;
        } else {
            previous_number = 0;
        }

        if index == 1 {
            is_ascending = *number > previous_number;
        } else {
            is_ascending = true;
        }

        if !is_safe_sequence(previous_number, *number, is_ascending) {
            return false;
        }

        previous_number = *number;
    }

    return true;
}

fn is_safe_sequence(previous_number: i64, current_number: i64, is_ascending: bool) -> bool {
    if previous_number == current_number {
        return false;
    }

    if is_ascending && previous_number > current_number || !is_ascending && previous_number < current_number {
        return false;
    }

    let difference: i64 = previous_number - current_number;
    
    if difference.abs() > 3 {
        return false;
    }

    return true;
}
